src/datamodel/quests/quest_new.rs

```diff
@@ -1,64 +1,45 @@
-use std::collections::HashMap;
-use std::rc::Rc;
+use std::hash::Hash;
 
-#[derive(Debug, PartialEq)]
+use crate::core::election::Election;
+
+#[derive(Hash, PartialEq, Eq, Clone, Debug, Copy)]
 pub enum Vote {
     Success,
     Failed,
 }
 
-pub struct QuestObjective {
-    pub num_players: u32,
-    pub required_successes: u32,
-}
-
-pub struct QuestNew {
-    members: HashMap<&'static str, bool>,
-    objective: Rc<QuestObjective>,
-    success_votes: u32,
+pub struct QuestNew<'a> {
+    election: Election<Vote>,
+    winner_rule: &'a (Vote, usize),
 }
 
-impl QuestNew {
-    pub fn new(members: Vec<&'static str>, objective: Rc<QuestObjective>) -> QuestNew {
-        let members: HashMap<&'static str, bool> =
-            members.into_iter().map(|m| (m, false)).collect();
-        QuestNew {
-            objective,
-            members,
-            success_votes: 0,
-        }
+impl<'a> QuestNew<'a> {
+    pub fn new(quest_member: &[String], winner_rule: &'a (Vote, usize)) -> QuestNew<'a> {
+        assert!(quest_member.len() >= winner_rule.1, "Winner's rule cannot be bigger than quest member number");
+        let election = Election::<Vote>::new(quest_member);
+        QuestNew { election, winner_rule, }
     }
 
-    pub fn add_vote(&mut self, player: &str, vote: Vote) {
-        if self.members[player] == true {
-            return;
-        }
-        if vote == Vote::Success {
-            self.success_votes += 1;
-        }
-        *self.members.get_mut(player).unwrap() = true;
+    pub fn vote(&mut self, quest_member: &String, vote: Vote) {
+        self.election.vote(quest_member, vote);
     }
 
-    pub fn finish_quest(self) -> Result<FinishedQuestNew, QuestNew> {
-        let unvoted: Vec<bool> = self.members.iter().map(|m| *m.1).filter(|v| !*v).collect();
-        if unvoted.len() > 0 {
-            Err(self)
-        } else {
-            if self.success_votes == self.objective.required_successes {
-                Ok(FinishedQuestNew { success: true })
-            } else {
-                Ok(FinishedQuestNew { success: false })
-            }
+    pub fn finish_quest(self) -> QuestResult {
```

Comment:
What happens now if some people didn't vote?

Reply:
There should also be a test for that case.

Reply:
That's tested in `election.rs`, and an `Err(Election)` is returned. `Election` enforces compulsory suffrage, so everyone must vote. 
I agree this is a flaw in our system and I think there are 2 ways to treat it:
- We stop returning a `Result` with `count_votes` in `election.rs` and we replace it with a `panic` turning it in an [unrecoverable error](https://doc.rust-lang.org/book/ch09-01-unrecoverable-errors-with-panic.html#unrecoverable-errors-with-panic). It might seem a little extreme but this approach has 1 up side: It is really simple to implement. I don't know how robust could be this solution, probably not much.
- We handle the error as [recoverable](https://doc.rust-lang.org/book/ch09-02-recoverable-errors-with-result.html#recoverable-errors-with-result) by the book, returning `Result<QuestResult, QuestNew>` and replacing `self.election.count_votes().ok().unwrap()` with a `match` that handles the return of `self.election.count_votes()`.

What do you think @julianmateu ? Blue pill or red pill?

Reply:
I've taken the red pill, so we are now handling errors with a `Result` type.

Reply:
Ok. I still need we should add a test here. We're testing the behaviour of the `Quest`, which will depend on behaviour of the `Election`, but it is behaviour that is part of the `Quest` contract/API, so it should be tested. For example, if we refactor `Quest` so that it doesn't depend on `Election` anymore, we still want to keep this behaviour.

Reply:
I completly agree, I forgot about the test that's why I haven't added. I'll do it tomorrow.

Reply:
I've just added it